### PR TITLE
Add a case to `lint_search_is_some` to handle searching strings

### PR DESCRIPTION
**Pull request by rsulli55**
Fixes: #6010 

This adds a lint which recommends using `contains()` instead of `find()` followed by `is_some()` on strings as suggested in #6010.

This was added as an additional case to
https://github.com/rust-lang/rust-clippy/blob/5af88e3c2d8cc4fb74a0e455381669930ee3a31a/clippy_lints/src/methods/mod.rs#L3037 

I would really appreciate any comments/suggestions for my code!

changelog: Added case to `lint_search_is_some` to handle searching strings

### CODE REVIEW

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (pickfire) [inline clippy_lints/src/find_is_some_on_strs.rs:50]: > on a string

I think we should also mention slice.

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (pickfire) [inline clippy_lints/src/find_is_some_on_strs.rs:68]: Should we return quickly if this passes rather then checking the next one?

File: tests/ui/find_is_some_on_strs.stderr
Comment (pickfire) [inline tests/ui/find_is_some_on_strs.stderr:5]: I think having two `:` looks weird, how about?
```
help: try `contains("world")`
```

**Comment by pickfire** [comment]
> I followed the tutorial and called cargo dev new_lint to make this lint, but after looking more through the code base, I am not sure if this instead should be place in clippy_lints/src/methods/mod.rs. Specfically, 

I think this should go into `src/methods/mod.rs` but I don't have much experience contributing to clippy myself, better wait for reviewer.

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (rsulli55) [reply clippy_lints/src/find_is_some_on_strs.rs:19]: Thanks!

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (rsulli55) [reply clippy_lints/src/find_is_some_on_strs.rs:50]: Thanks for the comment. I am still new to rust so I wasn't exactly sure what to call it. I was using string informally to mean that it could be a `String` or `&str` (or possibly `str`, but I'm not sure how one obtains a value of that type).

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (rsulli55) [reply clippy_lints/src/find_is_some_on_strs.rs:55]: Thanks.  I agree that `STRS` is not the optimal name.  I am new to rust, so is the word slice used to refer to any of the types `String`, `&str`, or `str`?

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (rsulli55) [reply clippy_lints/src/find_is_some_on_strs.rs:68]: Yes, that would be better.  I wasn't sure if `return` statements should be avoided or not.  But it sounds like using `return`s to exit early is good?

File: tests/ui/find_is_some_on_strs.stderr
Comment (rsulli55) [reply tests/ui/find_is_some_on_strs.stderr:5]: I agree, although I don't know if I have control over it.  The extra `:` is added by `span_lint_and_sugg()`.  I call it with `"try this"` for the corresponding argument (without `:`).

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (rsulli55) [reply clippy_lints/src/find_is_some_on_strs.rs:50]: Thanks!

File: clippy_lints/src/find_is_some_on_strs.rs
Comment (rsulli55) [reply clippy_lints/src/find_is_some_on_strs.rs]: I added the `return` statement, thanks for the suggestion!

**Comment by flip1995** [comment]
Yeah, your right, this should go into the methods module. IMO it doesn't even have to be a new lint, but a special case in the `SEARCH_IS_SOME` lint for `&str` and `String`.

**Comment by rsulli55** [comment]
Thanks for the comment!  I will move it to `lint_search_is_some()`

File: clippy_lints/src/methods/mod.rs
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:3117]:
```suggestion
                let mut applicability = Applicability::MachineApplicable;
                let find_arg = snippet_with_applicability(cx, search_args[1].span, "..", &mut applicability);
                span_lint_and_sugg(
                    cx,
                    SEARCH_IS_SOME,
                    method_span.with_hi(expr.span.hi()),
                    &msg,
                    "try this",
                    format!("contains({})", find_arg),
                    applicability,
                );
```

File: tests/ui/methods.rs
Comment (flip1995) [inline tests/ui/methods.rs:197]: It would be great if you could move those tests to a new file, named after this lint. After that you should be able to add `// run-rustfix` to the test file and simultaneously check if the suggestion is correct.

The `find().is_some()` part of the lint has cases where no suggestion is emitted. You can deal with this by splitting up the files further into `<lint_name>_fixable.rs` and `<lint_name>.rs`.

**Review by flip1995 — COMMENTED** [review]
> So, I have done that but unfortunately, the commits are not so nice now (one of them is mostly just deleting files). Sorry about that, I am still pretty new to the PR workflow.

Don't worry about this for now, we can deal with that right before merging.

---

impl LGTM overall. Just a NIT and splitting up the tests left to do.

File: clippy_lints/src/methods/mod.rs
Comment (rsulli55) [reply clippy_lints/src/methods/mod.rs:3117]: Thanks!

File: tests/ui/methods.rs
Comment (rsulli55) [reply tests/ui/methods.rs:197]: Ok, thanks I will get started on that.

Comment (rsulli55) [reply tests/ui/methods.rs:197]: Sorry, I should have clarified this earlier, but should I move only the new tests I added (the ones dealing with strings) to the separate file, or all the tests in `search_is_some()`?
https://github.com/rust-lang/rust-clippy/blob/3bd98895f13112912df3a7f23923470b07d53a09/tests/ui/methods.rs#L141

Comment (flip1995) [reply tests/ui/methods.rs:197]: All the tests for the `search_is_some` lint please. GH doesn't let you comment on unchanged code ;)

Comment (rsulli55) [reply tests/ui/methods.rs]: Thanks!

File: clippy_lints/src/methods/mod.rs
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:3107]:
```suggestion
                    `contains()`";
```

File: clippy_lints/src/methods/mod.rs
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:3114]:
```suggestion
                    msg,
```

**Review by flip1995 — APPROVED** [review]
Thanks! Only a NIT left.

File: clippy_lints/src/methods/mod.rs
Comment (rsulli55) [reply clippy_lints/src/methods/mod.rs:3107]: Thanks!

File: clippy_lints/src/methods/mod.rs
Comment (rsulli55) [reply clippy_lints/src/methods/mod.rs:3114]: Thanks!

File: clippy_lints/src/methods/mod.rs
Comment (rsulli55) [reply clippy_lints/src/methods/mod.rs]: @flip1995: Did you want to remove the period after `contains()`?  I think this is causing the test to fail and I will need to run `update-all-references.sh` again to update the `stderr` file if so.  

However, I just squashed commits on my local copy to make a rebase easier, so the commits will be changed if I do this (not sure if that's a problem or not).

**Comment by flip1995** [comment]
@rsulli55 You have to update the reference files. After that, we can merge this.

**Comment by rsulli55** [comment]
@flip1995 Thanks.  I have done that, however I was having trouble rebasing my local branch on the upstream branch so I squashed some of my earlier commits.  Is that okay? I have been holding off pushing my local branch because I wasn't sure if that would cause problems.

File: clippy_lints/src/methods/mod.rs
Comment (rsulli55) [inline clippy_lints/src/methods/mod.rs:3076]: @flip1995 I added back in the period after `contains()` to match the other cases of `search_is_some` (they all end with a period).  Let me know if you do not want that and I will remove it and re-run `update-all-references`.

By the way, thank you for you patience with me!  Sorry if my questions are all basic things.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:3076]: Oh they're also in other places? Can you remove them there too, please? Rust error/hint/suggestion messages are always without punctuation at the end and start lower-case.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:3076]: You could also remove the second part of this message here completely and replace the `"try this"` below with ``use `contains()` instead``.

Comment (rsulli55) [reply clippy_lints/src/methods/mod.rs]: @flip1995 Ok, I think they should be correctly updated.  I also cleaned up the other lints in `search_is_some` that suggested using `any()`.

File: tests/ui/search_is_some.rs
Comment (flip1995) [inline tests/ui/search_is_some.rs:31]: Can you either rename `foo` or allow `blacklisted_names` in this test?

File: clippy_lints/src/methods/mod.rs
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:3063]:
```suggestion
            "called `is_some()` after searching an `Iterator` with `{}`",
```

**Review by flip1995 — APPROVED** [review]
Two more NITs left. This should be the last.

File: clippy_lints/src/methods/mod.rs
Comment (rsulli55) [reply clippy_lints/src/methods/mod.rs:3063]: Thanks!

File: tests/ui/search_is_some.rs
Comment (rsulli55) [reply tests/ui/search_is_some.rs]: Thanks, I changed it to `falsepos`.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**
